### ISSUE TITLE
Feature Request: NMscanData() to give more info on error

### ISSUE THREAD

**Issue by billdenney** [philipdelff/NMdata#29]
I just got the error "After applying filters to input data, the resulting number of rows differ from the number of rows in output data"

It would help track down the source of the error if the number of rows in the input and output were reported.  (I realize that the NMdata-preferred solution is to use a ROWID column.  I'm trying to work within someone else's data management for the moment.)

My preferred error would look something like the following:

```
After applying filters to input data, the resulting number of rows differ (input = 123 rows) from the number of rows in output data (output = 456 rows)
```

**Comment by philipdelff** [comment]
Thanks Bill,

I can certainly add this information to the error message. 

Even though I encourage including a unique row identifier input and output, NMdata certainly supports not doing so. However, the combination of output and input then becomes dependent on NMdata interpreting the ACCEPT/IGNORE statements in the $DATA block of the control stream. I think either there is an issue with the input data or $INPUT, or the control stream outputs something in $TABLE that NMdata can't handle (eg, simulations where output events are multiples of input events are not supported for now).

One way you may want to debug this situation is to read the input data like NMscanData by running
inp <- NMscanInput("run.lst")
You can then manually inspect inp. You can also read the metadata info where you will see how the data was filtered by running:
NMinfo(inp)

I would also try to run NMcheckData on the control stream:
NMcheckData("run.lst")
That will read the data like NONMEM would and try to identify problems on the input data.

Let me know if any of this leads you to identifying the problem, and we can see if we should improve some checks to help identifying the problem faster.

**Comment by billdenney** [comment]
In the end, the error was real.  I was loading old model results with a new data file.  When I fixed updated to the new model results, everything worked as expected.

**Comment by philipdelff** [comment]
Glad to hear. This reminds me that the warnings related to timestamps are too long, and this may be why you didn't realize. NMscanData should have told you that the input data was newer than the output control stream unless you used check.time=F. Again, I want NMdata to help spot problems like these as fast as possible, so I am always open to ideas on checks and messages to include.

**Comment by billdenney** [comment]
It did give me that warning, but with my way of working (via a git repository with model estimation sometimes on separate machines from exploratory analysis and reporting), sometimes the file times are less logical than they would be if everything happened on the same system.

**Comment by philipdelff** [comment]
It is not uncommon to have R and Nonmem running on different systems as you are describing. Input data can also very well be generated on a different system than where the post-processing (and NMscanData) is run. 

I don't mind adding the number of rows to the error message. However, the relevant information is not how different they are but why they are different. Instead of using the file modification times (based on the file system) we could use 
- For output control stream: The time stamp included (first line AFAIK) in the output control stream 
- For input data: If NMwriteData was used, the time stamp saved in meta data. For csv files (when rds is not used by NMscanData), we need to make sure the time zone is recorded in the metadata. Right now, the csv meta data only includes the date-time without time zone which is insufficient for a reliable comparison. 

If we do it that way, the check would be useful in your case too, and you wouldn't have ignored it. Maybe we should then downplay (and not give warnings) when output tables are newer than input data. Currently, we also check if input control stream was modified (again, based on mtime) since Nonmem was run. I still think that check is important, but maybe not if $DATA is unchanged? $DATA is the only part of input control stream that NMscanData may use anyway, and that's only in case of a PSN setup (which is default) where the full path to the input data is unavailable in the output control stream.

Also for now, I will add better hints to the error message you got to what could cause this problem.

Thoughts?

**Comment by billdenney** [comment]
I think that date-time based checking will always be a tricky question to get right.

As a starting point, I'd use the time as UTC rather than including the time zone.  The reason to use UTC is that it is monotonic while time zones will have issues with overlapping times (when daylight savings time falls back there are duplicate hours unless the time zone is handled precisely-- which is not guaranteed).

For both the inputs and outputs, assuming that the times are synchronized between systems is often problematic.

I don't think that you need to get more complicated here in what you're testing.  I think that your warnings are useful, and while I ignored it, that's because of the way that git handles dates and times for files (which is also logical for the concept of how it works).

In summary, I think you're good with your current system.  But, the number of rows that was the mismatch would have gotten me to the underlying issue that I needed to update my local git repo a bit faster.

**Comment by philipdelff** [comment]
I added the dimensions to the error message as requested.

Also, the checks are now based on logged time stamps rather than file system "modification times". The logged time stamps are - i available:
- the time stamp recorded by Nonmem in the top of the output control stream
- the NMwriteData time stamp in metadata. This is included in .rds file, and for csv files found in the _meta.cvs file.
